sycl/source/detail/graph/graph_impl.hpp

```diff
@@ -510,6 +510,12 @@ class graph_impl : public std::enable_shared_from_this<graph_impl> {
   }
 
 private:
+  template <typename... Ts> node_impl &createNode(Ts &&...Args) {
+    MNodeStorage.push_back(
+        std::make_shared<node_impl>(std::forward<Ts>(Args)...));
+    return *MNodeStorage.back();
```

Comment:
Not sure if `node_impl` creation is under a mutex or not. If races are possible, then might need to change to 
```
   auto Ptr = make_shared();
   node_impl &Res = *Ptr;
   MNodeStorage.push_back(std::move(Ptr));
   return Res;
```

If that's the case, then `nodes_range` over `std::vector<node>` optimization needs to be examined for the race conditions as well.

Reply:
> Yes `node_impl` creation should always be under a mutex. Found a potential race and also gap in our e2e tests while investigating this: #19379

@reble , are you trying to investigate this more (or something else maybe) before continuing with the review of this PR?